compiler/vm/compilerbridge.nim

```diff
@@ -152,19 +154,43 @@ proc unpackResult(res: sink ExecutionResult; config: ConfigRef, node: PNode): PN
     result = config.newError(node, errKind, rId, instLoc(-1)):
                              [newIntNode(nkIntLit, ord(stId))]
 
-proc buildError(c: var TCtx, thread: VmThread, event: sink VMReport): ExecErrorReport  =
+proc createLegacyStackTrace(
+    c: TCtx, 
+    thread: VmThread, 
+    instLoc: InstantiationInfo = instLoc(-1)
+  ): VMReport =
+  let st = createStackTrace(c, thread)
+  result = VMReport(kind: rvmStackTrace,
+                    currentExceptionA: st.currentExceptionA,
+                    currentExceptionB: st.currentExceptionB,
+                    traceReason: vmEventToLegacyReportKind(st.traceReason),
+                    stacktrace: st.stacktrace,
+                    skipped: st.skipped,
+                    location: some source(c, thread),
+                    reportInst: toReportLineInfo(instLoc))
+
+proc buildError(c: TCtx, thread: VmThread, event: sink VMReport): ExecErrorReport  =
   ## Creates an `ExecErrorReport` with the `event` and a stack-trace for
   ## `thread`
-  result.stackTrace = createStackTrace(c, thread)
+  result.stackTrace = createLegacyStackTrace(c, thread, instLoc(-1))
   result.report = event
 
   if result.report.location.isNone():
     # use the location info of the failing instruction if none is provided
     result.report.location = some source(c, thread)
 
-  # set the stack-trace report information
-  result.stackTrace.location = some source(c, thread)
-  result.stackTrace.reportInst = toReportLineInfo(instLoc(-1))
+proc buildError(c: TCtx, thread: VmThread, event: sink VmEvent): ExecErrorReport  =
+  ## Creates an `ExecErrorReport` with the `event` and a stack-trace for
+  ## `thread`
+  result.stackTrace = createLegacyStackTrace(c, thread, instLoc(-1))
+  let location =
+    case event.kind
+    of vmEvtUserError:
+      event.srcLoc
+    else:
+      event.location.map(proc (l: PrgCtr): TLineInfo = c.debug[l])
+                    .get(source(c, thread))
```

Comment:
The `TCtx` is captured via a full copy here (which is very expensive, as `TCtx` can potentially own a very large amount of data).

Until the compiler is able to capture by-view and put the environment on the stack in cases like the one here, I think it's better to only use `map`, `flatMap`, etc. when the output only depends on the parameter.

Reply:
D'oh, I forgot about that, thanks. I'll stick with if/else.